File: Java/OpenGL-1.0(old_ver)/Loon-backend-Android/src/loon/core/timer/GameTime.java
/**
 * Copyright 2008 - 2012
 * 
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 * 
 * http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 * 
 * @project loon
 * @author cping
 * @email：javachenpeng@yahoo.com
 * @version 0.3.3
 */
package loon.core.timer;

public class GameTime {

	float _elapsedTime;
	float _totalTime;

	boolean _running;

	public GameTime() {
		_elapsedTime = _totalTime = 0f;
	}

	public GameTime(float totalGameTime, float elapsedGameTime) {
		_totalTime = totalGameTime;
		_elapsedTime = elapsedGameTime;
	}

	public GameTime(float totalRealTime, float elapsedRealTime,
			boolean isRunningSlowly) {
		_totalTime = totalRealTime;
		_elapsedTime = elapsedRealTime;
		_running = isRunningSlowly;
	}

	public void update(float elapsed) {
		_elapsedTime = elapsed;
		_totalTime += elapsed;
	}

	public void update(LTimerContext context) {
		update(context.getMilliseconds());
	}

	public void resetElapsedTime() {
		_elapsedTime = 0f;
	}

	public boolean isRunningSlowly() {
		return _running;
	}

	public float getMilliseconds() {
		return _elapsedTime * 1000;
	}

	public float getElapsedGameTime() {
		return _elapsedTime;
	}

	public float getTotalGameTime() {
		return _totalTime;
	}
	
}

File: C#/Loon2MonoGame/LoonMonoGame-Lib/loon/utils/timer/GameTime.cs
namespace loon.utils.timer
{

    public class GameTime
    {

        private static GameTime _instance = null;

        public static void FreeStatic()
        {
            _instance = null;
        }

        public static GameTime Get()
        {
            return GetInstance();
        }

        public static GameTime GetInstance()
        {
            if (_instance == null)
            {
                lock (typeof(GameTime))
                {
                    if (_instance == null)
                    {
                        _instance = new GameTime();
                    }
                }
            }
            return _instance;
        }

        float _elapsedTime;
        float _totalTime;

        bool _running;

        public static GameTime At()
        {
            return new GameTime();
        }

        public GameTime() : this(0f, 0f)
        {

        }

        public GameTime(float totalGameTime, float elapsedGameTime) : this(totalGameTime, elapsedGameTime, false)
        {

        }

        public GameTime(float totalRealTime, float elapsedRealTime, bool isRunningSlowly)
        {
            this._totalTime = totalRealTime;
            this._elapsedTime = elapsedRealTime;
            _running = isRunningSlowly;
        }

        public void Update(float elapsed)
        {
            this._elapsedTime = elapsed;
            this._totalTime += elapsed;
        }

        public void Update(LTimerContext context)
        {
            Update(context.GetMilliseconds());
        }

        public GameTime ResetElapsedTime()
        {
            _elapsedTime = 0f;
            return this;
        }

        public bool IsRunningSlowly()
        {
            return _running;
        }

        public float GetMilliseconds()
        {
            return MathUtils.Max(_elapsedTime * 1000, 10);
        }

        public float GetElapsedGameTime()
        {
            return _elapsedTime;
        }

        public float GetTotalGameTime()
        {
            return _totalTime;
        }


        public override string ToString()
        {
            StringKeyValue builder = new StringKeyValue("GameTime");
            builder.Kv("elapsedTime", _elapsedTime).Comma().Kv("totalTime", _totalTime).Comma().Kv("running", _running);
            return builder.ToString();
        }
    }
}
